streams/src/main/java/org/apache/kafka/streams/processor/internals/ImmutableRepartitionTopicConfig.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * repartition topic config which doesn't permit changing number of partitions
+ */
+public class ImmutableRepartitionTopicConfig extends RepartitionTopicConfig {
+    private final int numberOfPartitions;
+
+    ImmutableRepartitionTopicConfig(final String name,
+                                    final int numberOfPartitions,
+                                    final Map<String, String> topicConfigs) {
+        super(name, topicConfigs);
+        validateNumberOfPartitions(numberOfPartitions);
+        this.numberOfPartitions = numberOfPartitions;
+    }
+
+    @Override
+    public Optional<Integer> numberOfPartitions() {
+        return Optional.of(numberOfPartitions);
+    }
+
+    @Override
+    public void setNumberOfPartitions(final int numberOfPartitions) {
+        //no-op
```

Comment:
Should be better throw here as this should never be called?

What raised the question (we you actually do check the type already: should be flip the hierarchy as being mutable is a superset of being immutable and thus it should be `RepartitionTopicConfig extends ImmutableRepartitionTopicConfig` -- for this case, `ImmutableRepartitionTopicConfig` would not have this method at all what seems to be cleaner)

Reply:
Originally I've implemented this method with throwing an exception. But it seems like this method is being called from various places, like `StreamsPartitionAssignor#prepareTopic`. My thinking was - instead of each individual caller checking if `InternalTopicConfig` supports setting number of partitions, it makes more sense to delegate this to the "builder" that chooses appropriate implementation based on some logic.

`setNumberOfPartitions` is part of `InternalTopicConfig` so even i flip the hierarchy, I can't avoid `setNumberOfPartitions` method. So I don't think it gives any benefit if we flip the hierarchy. I also thought instead of adding new class, I could maybe enhance `RepartitionTopicConfig` to support "immutability" in cases when topics are for `repartition` operation, but in that case I have to add a flag to the class to indicate that this `RepartitionTopicConfig` is actually for `repartition` operation which seems a bit worse compared to introducing new class. Thoughts?

Reply:
This class was removed. Check my comment here: https://github.com/apache/kafka/pull/7170#discussion_r373849898